### PR TITLE
Add logging support from Winston with colored output

### PR DESCRIPTION
**Pull request by rizwanmustafa**
The logs can help in production and the colored output will be more readable.

From this:
Non Colored Output

To This:
Logger Output

### CODE REVIEW

**Comment by rizwanmustafa** [comment]
@Bruception A review please.

File: backend/init/redis.ts [hunk 1]
```diff
@@ -29,8 +33,9 @@ class RedisClient {
       await this.connection.connect();
       this.connected = true;
     } catch (error) {
-      console.error(error.message);
+      logger.error(`Redis - ${error.message}`);
       console.error(
```
Comment (Bruception) [inline backend/init/redis.ts:37]: Change this to use logger

File: backend/api/controllers/result.ts [hunk 2]
```diff
@@ -4,7 +4,7 @@ import PublicStatsDAO from "../../dao/public-stats";
 import BotDAO from "../../dao/bot";
 import { roundTo2, stdDev } from "../../utils/misc";
 import objectHash from "object-hash";
-import Logger from "../../utils/logger";
+import { logToDb } from "../../utils/logger";
```
Comment (Bruception) [inline backend/api/controllers/result.ts:7]: Condense import,

`import { logToDb, logger }`

File: backend/api/controllers/result.ts [hunk 3]
```diff
@@ -17,19 +17,20 @@ import {
...
 
 try {
   if (anticheatImplemented() === false) throw new Error("undefined");
-  console.log("Anticheat module loaded");
+  logger.log("success", "Anticheat - Anticheat module loaded");
 } catch (e) {
   if (process.env.MODE === "dev") {
-    console.error(
-      "No anticheat module found. Continuing in dev mode, results will not be validated."
+    logger.log(
```
Comment (Bruception) [inline backend/api/controllers/result.ts:27]: Should we warning

File: backend/api/controllers/result.ts [hunk 4]
```diff
@@ -108,14 +109,15 @@ class ResultController {
       if (process.env.MODE !== "dev") {
         throw new Error("No anticheat module found");
       }
-      console.error(
-        "No anticheat module found. Continuing in dev mode, results will not be validated."
+      logger.log(
```
Comment (Bruception) [inline backend/api/controllers/result.ts:112]: Why not: `logger.error`?

File: backend/api/controllers/result.ts [hunk 5]
```diff
@@ -210,8 +212,9 @@ class ResultController {
         if (process.env.MODE !== "dev") {
           throw new Error("No anticheat module found");
         }
-        console.error(
-          "No anticheat module found. Continuing in dev mode, results will not be validated."
+        logger.log(
```
Comment (Bruception) [inline backend/api/controllers/result.ts:215]: Same as above

File: backend/init/db.ts [hunk 6]
```diff
@@ -53,9 +54,10 @@ class DatabaseClient {
       this.db = this.mongoClient.db(DB_NAME);
       this.connected = true;
     } catch (error) {
-      console.error(error.message);
-      console.error(
-        "Failed to connect to database. Exiting with exit status code 1."
+      logger.error(`Database - ${error.message}`);
+      logger.log(
```
Comment (Bruception) [inline backend/init/db.ts:58]: Same as above

File: backend/init/redis.ts [hunk 7]
```diff
@@ -13,7 +14,10 @@ class RedisClient {
 
     if (!REDIS_URI) {
       if (process.env.MODE === "dev") {
-        console.log("No redis configuration provided. Running without redis.");
+        logger.log(
```
Comment (Bruception) [inline backend/init/redis.ts:17]: same

File: backend/middlewares/auth.ts [hunk 8]
```diff
@@ -123,9 +124,7 @@ async function authenticateWithBearerToken(
       email: decodedToken.email ?? "",
     };
   } catch (error) {
-    console.log("-----------");
-    console.log(error.errorInfo.code);
-    console.log("-----------");
+    logger.error(`General - ${error.errorInfo.code.toString()}`);
```
Comment (Bruception) [inline backend/middlewares/auth.ts:127]: General -> Firebase Auth Error

File: backend/middlewares/error.ts [hunk 9]
```diff
@@ -1,9 +1,10 @@
 import db from "../init/db";
 import { v4 as uuidv4 } from "uuid";
-import Logger from "../utils/logger";
+import { logToDb } from "../utils/logger";
```
Comment (Bruception) [inline backend/middlewares/error.ts:3]: Condense imports

File: backend/server.ts [hunk 10]
```diff
@@ -11,44 +11,49 @@ import { version } from "./version";
...
 import RedisClient from "./init/redis";
 import George from "./tasks/george";
+import { logger } from "./utils/logger";
 
 async function bootServer(port: number): Promise<Server> {
   try {
     console.log(`Connecting to database ${process.env.DB_NAME}...`);
     await db.connect();
-    console.log("Connected to database");
+    logger.log("success", "Database - Connected to database");
```
Comment (Bruception) [inline backend/server.ts:20]: This is repetitive consider: `logger.success`

File: backend/server.ts [hunk 11]
```diff
@@ -11,44 +11,49 @@ import { version } from "./version";
...
 
     recordServerVersion(version);
   } catch (error) {
-    console.error("Failed to boot server");
-    console.error(error);
+    logger.error("Server - Failed to boot server");
+    if (typeof error === "string" || error instanceof String) {
+      logger.error(`Server - ${error}`);
+    } else {
+      console.error(error);
```
Comment (Bruception) [inline backend/server.ts:55]: why console log?

File: backend/server.ts [hunk 12]
```diff
@@ -11,44 +11,49 @@ import { version } from "./version";
...
     jobs.forEach((job) => job.start());
-    console.log("Cron jobs started");
+    logger.log("success", "Cron Jobs - Cron jobs started");
 
     recordServerVersion(version);
   } catch (error) {
-    console.error("Failed to boot server");
-    console.error(error);
+    logger.error("Server - Failed to boot server");
+    if (typeof error === "string" || error instanceof String) {
```
Comment (Bruception) [inline backend/server.ts:52]: Why this check?

File: backend/utils/logger.ts [hunk 13]
```diff
@@ -1,22 +1,99 @@
...
+const customLevels = {
+  error: 0,
+  warning: 1,
+  info: 2,
+  success: 3,
+};
+
+const timestampFormat = format.timestamp({
+  format: "MMM-DD-YYYY HH:mm:ss",
+});
```
Comment (Bruception) [inline backend/utils/logger.ts:32]: Strange spacing here. Add new lines here and below

File: backend/middlewares/error.ts [hunk 14]
```diff
@@ -52,12 +53,15 @@ async function errorHandlingMiddleware(
         endpoint: req.originalUrl,
       });
     } catch (e) {
-      console.error("Failed to save error.");
-      console.error(e);
+      logger.error("General - Failed to save error.");
+      if (typeof e === "string" || e instanceof String) {
```
Comment (Bruception) [inline backend/middlewares/error.ts:57]: Why this check?

File: backend/utils/logger.ts [hunk 15]
```diff
@@ -1,22 +1,99 @@
+import "dotenv/config";
```
Comment (Bruception) [inline backend/utils/logger.ts:1]: Remove this import. It is not needed.

File: backend/init/configuration.ts [hunk 16]
```diff
@@ -50,7 +50,7 @@ class ConfigurationClient {
       attemptCacheUpdate &&
       this.lastFetchTime < Date.now() - CONFIG_UPDATE_INTERVAL
     ) {
-      console.log("Cached configuration is stale.");
+      Logger.info("General - Cached configuration is stale.");
```
Comment (Miodec) [inline backend/init/configuration.ts:53]: General?

File: backend/init/redis.ts [hunk 17]
```diff
@@ -29,8 +32,8 @@ class RedisClient {
       await this.connection.connect();
       this.connected = true;
     } catch (error) {
-      console.error(error.message);
-      console.error(
+      Logger.error(`${error.message}`);
```
Comment (Miodec) [inline backend/init/redis.ts:35]: Unnecessary template literal

File: backend/jobs/delete-old-logs.ts [hunk 18]
```diff
@@ -1,6 +1,6 @@
 import { CronJob } from "cron";
 import db from "../init/db";
-import Logger from "../utils/logger";
+import { Logger } from "../utils/logger";
```
Comment (Miodec) [inline backend/jobs/delete-old-logs.ts:3]: Why is this import destructured while others are not?

**Review by Miodec — CHANGES_REQUESTED** [review]

File: backend/middlewares/error.ts [hunk 19]
```diff
@@ -52,12 +52,11 @@ async function errorHandlingMiddleware(
...
     } catch (e) {
-      console.error("Failed to save error.");
-      console.error(e);
+      Logger.error("Logging to db failed.");
+      Logger.error(e);
     }
   } else {
-    console.error(error.message);
-    console.error(error.stack);
+    Logger.error(`General - Error: ${error.message} Stack: ${error.stack}`);
```
Comment (Miodec) [inline backend/middlewares/error.ts:59]: Remove general

**Review by Miodec — CHANGES_REQUESTED** [review]

File: backend/example.env [hunk 20]
```diff
@@ -1,6 +1,9 @@
 DB_NAME=monkeytype
 DB_URI=mongodb://localhost:27017
 REDIS_URI=redis://localhost:6379
+LOG_FOLDER_PATH=./logs/
+LOG_FILE_MAX_SIZE=10485760
+# Default log file max sizee is 10 MB
```
Comment (Bruception) [inline backend/example.env:6]: Move this comment up. Also add ./logs to gitignore

File: backend/utils/logger.ts [hunk 21]
```diff
@@ -1,22 +1,111 @@
...
+});
+
+const logToDb = async (
+  event: string,
+  message: any,
+  uid?: string
+): Promise<void> => {
+  const logsCollection = db.collection<Log>("logs");
+
+  console.log(new Date(), "\t", event, "\t", uid, "\t", message);
```
Comment (Bruception) [inline backend/utils/logger.ts:94]: Maybe we can convert this log to use logger

File: backend/example.env (hunk 20 above)
Comment (Miodec) [reply backend/example.env:6]: Moved. Logs is already in gitignore

File: backend/example.env [hunk 22]
```diff
@@ -2,8 +2,8 @@ DB_NAME=monkeytype
 DB_URI=mongodb://localhost:27017
 REDIS_URI=redis://localhost:6379
 LOG_FOLDER_PATH=./logs/
-LOG_FILE_MAX_SIZE=10485760
```
Comment (Bruception) [inline backend/example.env]: Typo here sizee

**Review by Bruception — APPROVED** [review]
LGTM IIW

**Review by Bruception — APPROVED** [review]
LGTM IIW

**Comment by CameronCT** [comment]
LCOSTBIT

**Pull request merged**